### PR TITLE
Fix: `Simplify` handle all types

### PR DESCRIPTION
**Pull request by skarab42**
Related to https://github.com/sindresorhus/type-fest/pull/408#discussion_r909523586

- Fix the case where the type to be simplified is a function. In this case we return the type without modification.

### CODE REVIEW

File: test-d/simplify.ts [hunk 1]
```diff
@@ -42,3 +42,25 @@ expectType<SomeInterfaceAsTypeWrittenByHand>(valueAsInterface);
 expectAssignable<Record<string, unknown>>(valueAsLiteral);
 expectAssignable<Record<string, unknown>>(valueAsSimplifiedInterface);
 expectNotAssignable<Record<string, unknown>>(valueAsInterface); // Index signature is missing in interface
+
+// Should return the original type if it is not simplifiable, like a function.
+type SomeFunction = (type: string) => string;
+expectType<Simplify<SomeFunction>>((type: string) => type);
+
+// Sould simplify to public interface (what it is currently doing)? Return the Class type? Or something else?
```
Comment (skarab42) [inline test-d/simplify.ts:50]: I also wonder about classes. Currently it returns a simplified public version of the class interface.
What do you expect? the public interface? complete interface? or just the original type?

Comment (sindresorhus) [reply test-d/simplify.ts:50]: I would say, just the original type. The point of the utility is to simplify nested and complicated types.

**Comment by skarab42** [comment]
@sindresorhus I think I have found a trick to support any type without testing all cases. The strategy is if the simplified type extends the source type then we return the simplified type otherwise the source.

I also added the `deep` option to apply the simplification recursively.

File: source/simplify.d.ts [hunk 2]
```diff
@@ -1,3 +1,24 @@
+import {Class, Constructor} from './basic';
+
+/**
+Simplify options.
+
+@param deep - Do the simplification recursively (default: false).
```
Comment (sindresorhus) [inline source/simplify.d.ts:6]: This should be attached to the `deep` property itself. And use `@default` for the default value

File: source/simplify.d.ts [hunk 3]
```diff
@@ -55,4 +76,9 @@ fn(someInterface as Simplify<SomeInterface>); // Good: transform an `interface`
 
 @category Object
 */
-export type Simplify<T> = {[KeyType in keyof T]: T[KeyType]};
+export type Simplify<
+	AnyType,
+	Options extends SimplifyOptions = {},
+> = Flatten<AnyType> extends AnyType
```
Comment (sindresorhus) [inline source/simplify.d.ts:82]:
```suggestion
> = Flatten<AnyType, Options> extends AnyType
```

?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/simplify.d.ts (hunk 3 above)
Comment (skarab42) [reply source/simplify.d.ts:82]: No need to pass the options here, it's just for the test, if it doesn't pass without the deep there's no point in going any further

**Pull request merged**